activemodel/lib/active_model/serialization.rb

```diff
@@ -120,6 +120,13 @@ def serializable_hash(options = nil)
       hash
     end
 
+    # Hook method for defining the set of available attribute names. Override
+    # this method in subclasses where, e.g., it's possible to know the names
+    # without having to make an attributes hash.
+    def attribute_names
+      attributes.keys
+    end
```

Comment:
Interestingly, [I just implemented something like this in ActiveModelSerializers::Model](https://github.com/rails-api/active_model_serializers/blob/0422a1e7724023aa4c54b7fee683660641d7d523/lib/active_model_serializers/model.rb#L9-L43) to use attribute_names, and ran into a bug where my 'attribute_names' was conflicting with the one in ActiveModel::Serialization (which led me here).  Also, I got the idea from Crafting Rails Applications by Jose Valim, one of the [original authors of this part of the code and AMS](https://github.com/rails-api/active_model_serializers/blob/0422a1e7724023aa4c54b7fee683660641d7d523/CHANGELOG.md#prehistory). So, good provenance.